numpy/f2py/cfuncs.py

```diff
@@ -1152,6 +1177,30 @@
 """
 
 
+cfuncs['try_pyarr_from_character'] = """\
+static int try_pyarr_from_character(PyObject* obj, character* v) {
+    PyArrayObject *arr = (PyArrayObject*)obj;
+    if (!obj) return -2;
+    if (PyArray_Check(obj)) {
+        if (PyArray_DESCR(arr)->type==NPY_CHARLTR || PyArray_DESCR(arr)->type==NPY_STRINGLTR)  {
```

Comment:
Honestly, I think probably NumPy assumes that `NPY_CHAR` is basically never used as a type-number in a few places by now.
For `PyArray_DESCR(arr)->num == NPY_STRING` is guaranteed to be correct unless someone had the audacity to set `->num` manually.  We can assume this does not happen.

Reply:
This was my early change that turns out to be suboptimal. Now using
```c
if (F2PY_ARRAY_IS_CHARACTER_COMPATIBLE(arr)) ...
```

Reply:
Fixed.